types/account.go

```diff
@@ -3,97 +3,56 @@ package types
-	"fmt"
 
-	"gopkg.in/yaml.v2"
+	yaml "gopkg.in/yaml.v2"
 
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
-	"github.com/cosmos/cosmos-sdk/x/auth"
-	"github.com/cosmos/cosmos-sdk/x/auth/exported"
 	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
 
 	ethcmn "github.com/ethereum/go-ethereum/common"
 	ethcrypto "github.com/ethereum/go-ethereum/crypto"
 )
 
-var _ exported.Account = (*EthAccount)(nil)
-var _ exported.GenesisAccount = (*EthAccount)(nil)
-
-func init() {
-	authtypes.RegisterAccountTypeCodec(&EthAccount{}, EthAccountName)
-}
+var (
+	_ authtypes.AccountI                 = (*EthAccount)(nil)
+	_ authtypes.GenesisAccount           = (*EthAccount)(nil)
+	_ codectypes.UnpackInterfacesMessage = (*EthAccount)(nil)
+)
 
 // ----------------------------------------------------------------------------
 // Main Ethermint account
 // ----------------------------------------------------------------------------
 
-// EthAccount implements the auth.Account interface and embeds an
-// auth.BaseAccount type. It is compatible with the auth.AccountKeeper.
-type EthAccount struct {
-	*authtypes.BaseAccount `json:"base_account" yaml:"base_account"`
-	CodeHash               []byte `json:"code_hash" yaml:"code_hash"`
-}
-
 // ProtoAccount defines the prototype function for BaseAccount used for an
 // AccountKeeper.
-func ProtoAccount() exported.Account {
+func ProtoAccount() authtypes.AccountI {
 	return &EthAccount{
-		BaseAccount: &auth.BaseAccount{},
+		BaseAccount: &authtypes.BaseAccount{},
 		CodeHash:    ethcrypto.Keccak256(nil),
 	}
 }
 
 // EthAddress returns the account address ethereum format.
 func (acc EthAccount) EthAddress() ethcmn.Address {
-	return ethcmn.BytesToAddress(acc.Address.Bytes())
-}
-
-// TODO: remove on SDK v0.40
-
-// Balance returns the balance of an account.
-func (acc EthAccount) Balance(denom string) sdk.Int {
-	return acc.GetCoins().AmountOf(denom)
-}
-
-// SetBalance sets an account's balance of the given coin denomination.
-//
-// CONTRACT: assumes the denomination is valid.
-func (acc *EthAccount) SetBalance(denom string, amt sdk.Int) {
-	coins := acc.GetCoins()
-	diff := amt.Sub(coins.AmountOf(denom))
-	switch {
-	case diff.IsPositive():
-		// Increase coins to amount
-		coins = coins.Add(sdk.NewCoin(denom, diff))
-	case diff.IsNegative():
-		// Decrease coins to amount
-		coins = coins.Sub(sdk.NewCoins(sdk.NewCoin(denom, diff.Neg())))
-	default:
-		return
-	}
-
-	if err := acc.SetCoins(coins); err != nil {
-		panic(fmt.Errorf("could not set %s coins for address %s: %w", denom, acc.EthAddress().String(), err))
-	}
+	return ethcmn.BytesToAddress(acc.GetAddress().Bytes())
 }
 
 type ethermintAccountPretty struct {
-	Address       sdk.AccAddress `json:"address" yaml:"address"`
-	EthAddress    string         `json:"eth_address" yaml:"eth_address"`
-	Coins         sdk.Coins      `json:"coins" yaml:"coins"`
```

Comment:
just confirming, the coin is now stored in the bank module and not on the account,?

Reply:
yup